File: libs/monoboard/src/main/kotlin/mono/graphics/board/PainterBoard.kt
package mono.graphics.board

import mono.common.Characters.TRANSPARENT_CHAR
import mono.graphics.bitmap.MonoBitmap
import mono.graphics.geo.Point
import mono.graphics.geo.Rect
import mono.graphics.geo.Size

/**
 * A model class to manage drawn pixel.
 * This is where a pixel is represented with its absolute position.
 */
internal class PainterBoard(internal val bound: Rect) {
    private val validColumnRange = 0 until bound.width
    private val validRowRange = 0 until bound.height

    private val matrix: List<List<Pixel>> = List(bound.height) {
        List(bound.width) { Pixel() }
    }

    fun clear() {
        for (row in matrix) {
            for (cell in row) {
                cell.reset()
            }
        }
    }

    /**
     * Fills with another [PainterBoard].
     * If a pixel in input [PainterBoard] is transparent, the value in the current board at that
     * position won't be overwritten.
     */
    fun fill(board: PainterBoard) {
        val position = board.bound.position
        val inMatrix = board.matrix

        if (matrix.isEmpty() || matrix.first().isEmpty()) {
            return
        }
        val inMatrixBound = Rect(position, Size(inMatrix.first().size, inMatrix.size))

        val overlap = bound.getOverlappedRect(inMatrixBound) ?: return
        val (startCol, startRow) = overlap.position - bound.position
        val (inStartCol, inStartRow) = overlap.position - position

        for (r in 0 until overlap.height) {
            val src = inMatrix[inStartRow + r]
            val dest = matrix[startRow + r]

            src.subList(inStartCol, inStartCol + overlap.width).forEachIndexed { index, pixel ->
                if (!pixel.isTransparent) {
                    dest[startCol + index].set(pixel.char, pixel.highlight)
                }
            }
        }
    }

    /**
     * Fills with a bitmap and the highlight state of that bitmap from [position].
     * If a pixel in input [bitmap] is transparent, the value in the current board at that
     * position won't be overwritten.
     */
    fun fill(position: Point, bitmap: MonoBitmap, highlight: Highlight) {
        if (bitmap.isEmpty()) {
            return
        }
        val inMatrix = bitmap.matrix

        val inMatrixBound = Rect(position, bitmap.size)

        val overlap = bound.getOverlappedRect(inMatrixBound) ?: return
        val (startCol, startRow) = overlap.position - bound.position
        val (inStartCol, inStartRow) = overlap.position - position

        for (r in 0 until overlap.height) {
            val src = inMatrix[inStartRow + r]
            val dest = matrix[startRow + r]

            src.forEachIndex(inStartCol, inStartCol + overlap.width) { index, char ->
                dest[startCol + index].set(char, highlight)
            }
        }
    }

    /**
     * Force values overlap with [rect] to be [char] regardless they are [TRANSPARENT_CHAR].
     *
     * Note: This method is for testing only
     */
    fun fill(rect: Rect, char: Char, highlight: Highlight) {
        val overlap = bound.getOverlappedRect(rect) ?: return
        val (startCol, startRow) = overlap.position - bound.position

        for (r in 0 until overlap.height) {
            val row = matrix[r + startRow]
            for (c in 0 until overlap.width) {
                row[c + startCol].set(char, highlight)
            }
        }
    }

    /**
     * Force value at [position] to be [char] with [highlight].
     *
     * Note: This method is for testing only
     */
    fun set(position: Point, char: Char, highlight: Highlight) =
        set(position.left, position.top, char, highlight)

    // This method is for testing only
    fun set(left: Int, top: Int, char: Char, highlight: Highlight) {
        val columnIndex = left - bound.left
        val rowIndex = top - bound.top
        if (columnIndex !in validColumnRange || rowIndex !in validRowRange) {
            return
        }
        matrix[rowIndex][columnIndex].set(char, highlight)
    }

    operator fun get(position: Point): Pixel? = get(position.left, position.top)

    fun get(left: Int, top: Int): Pixel? {
        val columnIndex = left - bound.left
        val rowIndex = top - bound.top
        return matrix.getOrNull(rowIndex)?.getOrNull(columnIndex)
    }

    override fun toString(): String =
        matrix.joinToString("\n", transform = ::toRowString)

    private fun toRowString(chars: List<Pixel>): String =
        chars.joinToString("") { it.toString() }
}

File: monosketch-svelte/src/lib/mono/monobitmap/board/painter-board.ts
import { HighlightType, Pixel } from '$mono/monobitmap/board/pixel';
import { Rect } from '$libs/graphics-geo/rect';
import { Size } from '$libs/graphics-geo/size';
import type { Point } from '$libs/graphics-geo/point';
import { MonoBitmap } from '$mono/monobitmap/bitmap/monobitmap';
import { isHalfTransparentChar } from '$mono/common/character';
import type { Char } from '$libs/char';
import type { CrossPoint } from '$mono/monobitmap/board/cross-point';
import { ListExt } from '$libs/sequence';
import list = ListExt.list;
import { CrossingResources } from '$mono/monobitmap/board/crossing-resources';
import isConnectableChar = CrossingResources.isConnectableChar;

/**
 * A model class to manage drawn pixel.
 * This is where a pixel is represented with its absolute position.
 */
export class PainterBoard {
    private readonly matrix: Pixel[][];

    constructor(private bound: Rect) {
        this.matrix = list(bound.height, () => list(bound.width, () => new Pixel()));
    }

    getBound = (): Rect => this.bound;

    clear = () => {
        for (const row of this.matrix) {
            for (const pixel of row) {
                pixel.reset();
            }
        }
    };

    /**
     * Fills with another [PainterBoard].
     * If a pixel in input [PainterBoard] is transparent, the value in the current board at that
     * position won't be overwritten.
     */
    fill(board: PainterBoard) {
        if (this.matrix.length === 0 || this.matrix[0].length === 0) {
            return;
        }

        const position = board.bound.position;
        const inMatrix = board.matrix;

        const inMatrixBound = new Rect(position, Size.of(inMatrix[0].length, inMatrix.length));

        const overlap = this.bound.getOverlappedRect(inMatrixBound);
        if (!overlap) {
            return;
        }
        const { left: startCol, top: startRow } = overlap.position.minus(this.bound.position);
        const { left: inStartCol, top: inStartRow } = overlap.position.minus(position);

        for (let r = 0; r < overlap.height; r++) {
            const src = inMatrix[inStartRow + r];
            const dest = this.matrix[startRow + r];

            for (let c = 0; c < overlap.width; c++) {
                const px = src[inStartCol + c];
                if (!px.isTransparent) {
                    dest[startCol + c].set(px.visualChar, px.directionChar, px.highlight);
                }
            }
        }
    };

    /**
     * Fills with a bitmap and the highlight state of that bitmap from [position] excepts crossing
     * points. Connection point are the point having the char which is connectable
     * ([CrossingResources.isConnectable]) and there is a character drawn at the position.
     * A list of [CrossPoint] will be returned to let [MonoBoard] able to adjust and draw the
     * adjusted character of the connection points.
     *
     * The main reason why it is required to let [MonoBoard] draws the connection points is the
     * painter board cannot see the pixel outside its bound which is required to identify the final
     * connection character.
     *
     * If a pixel in input [bitmap] is transparent, the value in the current board at that
     * position won't be overwritten.
     */
    fillBitmap(
        position: Point,
        bitmap: MonoBitmap.Bitmap,
        highlight: HighlightType,
    ): Array<CrossPoint> {
        if (bitmap.isEmpty()) {
            return [];
        }
        const inMatrix = bitmap.matrix;
        const inMatrixBound = new Rect(position, bitmap.size);

        const overlap = this.bound.getOverlappedRect(inMatrixBound);
        if (!overlap) {
            return [];
        }

        const { left: startCol, top: startRow } = overlap.position.minus(this.bound.position);
        const { left: inStartCol, top: inStartRow } = overlap.position.minus(position);
        const { left: boundColumn, top: boundRow } = this.bound.position;

        const crossPoints: CrossPoint[] = [];

        for (let r = 0; r < overlap.height; r++) {
            const bitmapRow = inStartRow + r;
            const painterRow = startRow + r;
            const src = inMatrix[bitmapRow];
            const dest = this.matrix[painterRow];

            for (const cell of src.asSequence(inStartCol, inStartCol + overlap.width)) {
                const index = cell.index - inStartCol;
                const bitmapColumn = inStartCol + index;
                const painterColumn = startCol + index;
                const pixel = dest[painterColumn];

                if (this.isApplicable(pixel, cell.visual)) {
                    // Not drawing half transparent character
                    // (full transparent character is removed by bitmap)
                    if (!isHalfTransparentChar(cell.visual)) {
                        pixel.set(cell.visual, cell.direction, highlight);
                    }
                } else {
                    // Crossing points will be drawn after finishing drawing all pixels of the
                    // bitmap on the Mono Board. Each unit painter board does not have enough
                    // information to decide the value of the crossing point.
                    crossPoints.push({
                        boardRow: painterRow + boundRow,
                        boardColumn: painterColumn + boundColumn,
                        visualChar: cell.visual,
                        directionChar: cell.direction,
                        leftChar: bitmap.getDirection(bitmapRow, bitmapColumn - 1),
                        rightChar: bitmap.getDirection(bitmapRow, bitmapColumn + 1),
                        topChar: bitmap.getDirection(bitmapRow - 1, bitmapColumn),
                        bottomChar: bitmap.getDirection(bitmapRow + 1, bitmapColumn),
                    });
                }
            }
        }

        return crossPoints;
    };

    /**
     * Force values overlap with [rect] to be [char] regardless they are [TRANSPARENT_CHAR].
     *
     * Note: This method is for testing only
     */
    fillRect = (rect: Rect, char: Char, highlight: HighlightType) => {
        const overlap = this.bound.getOverlappedRect(rect);
        if (!overlap) {
            return;
        }
        const { left: startCol, top: startRow } = overlap.position.minus(this.bound.position);

        for (let r = 0; r < overlap.height; r++) {
            const row = this.matrix[startRow + r];
            for (let c = 0; c < overlap.width; c++) {
                const pixel = row[startCol + c];
                pixel.set(char, char, highlight);
            }
        }
    };

    /**
     * Force value at [position] to be [char] with [highlight].
     *
     * Note: This method is for testing only
     */
    setPoint = (position: Point, char: Char, highlight: HighlightType) => {
        const columnIndex = position.left - this.bound.left;
        const rowIndex = position.top - this.bound.top;

        if (
            columnIndex < 0 ||
            columnIndex >= this.bound.width ||
            rowIndex < 0 ||
            rowIndex >= this.bound.height
        ) {
            return;
        }
        this.matrix[rowIndex][columnIndex].set(char, char, highlight);
    };

    get = (left: number, top: number): Pixel | null => {
        const columnIndex = left - this.bound.left;
        const rowIndex = top - this.bound.top;
        if (
            columnIndex < 0 ||
            columnIndex >= this.bound.width ||
            rowIndex < 0 ||
            rowIndex >= this.bound.height
        ) {
            return null;
        }
        return this.matrix[rowIndex][columnIndex];
    };

    private isApplicable(currentPixel: Pixel, toBeVisual: Char): boolean {
        if (currentPixel.isTransparent) {
            return true;
        }
        if (currentPixel.visualChar === toBeVisual) {
            return true;
        }
        return !isConnectableChar(toBeVisual);
    };

    toString = (): string => {
        return this.matrix.map(this.toRowString).join('\n');
    };

    private toRowString(row: Pixel[]): string {
        return row.map((pixel) => pixel.toString()).join('');
    }
}
